src/Microsoft.ML.Core/Data/Schema.cs

```diff
@@ -99,53 +115,106 @@ public sealed class Column
             /// </summary>
             public Metadata Metadata { get; }
 
-            public Column(string name, ColumnType type, Metadata metadata)
+            internal Column(string name, int index, bool isHidden, ColumnType type, Metadata metadata)
             {
-                Contracts.CheckNonEmpty(name, nameof(name));
-                Contracts.CheckValue(type, nameof(type));
-                Contracts.CheckValueOrNull(metadata);
+                Contracts.AssertNonEmpty(name);
+                Contracts.Assert(index >= 0);
+                Contracts.AssertValue(type);
+                Contracts.AssertValueOrNull(metadata);
 
                 Name = name;
+                Index = index;
+                IsHidden = isHidden;
                 Type = type;
-                Metadata = metadata;
+                Metadata = metadata ?? Metadata.Empty;
             }
 
             public override string ToString()
             {
-                var metadataString = (Metadata == null || Metadata.Schema.ColumnCount == 0) ?
-                    null : $" {{{string.Join(", ", Metadata.Schema.GetColumns().Select(x => x.column.Name))}}}";
+                var metadataString = (Metadata == null || Metadata.Schema.Count == 0) ?
+                    null : $" {{{string.Join(", ", Metadata.Schema.Select(x => x.Name))}}}";
                 return $"{Name}: {Type}{metadataString}";
             }
         }
 
+        /// <summary>
+        /// This class represents the schema of one column of a data view, without an attachment to a particular <see cref="Schema"/>.
+        /// </summary>
+        public sealed class DetachedColumn
```

Comment:
>DetachedColumn [](start = 28, length = 14)

I don't think we ought to do it right now, but the word "detached" is a little unfortunate since it suggests that it is something that has gone a "detach" operation, that is, it was previously attached, which is not how things often happen. But it may be fine. We'll see if anyone gets confused maybe, since I cannot think of a better word.